src/transport/SecurePairingSession.cpp

```diff
@@ -218,68 +280,242 @@ CHIP_ERROR SecurePairingSession::AttachHeaderAndSend(uint8_t msgType, System::Pa
     SuccessOrExit(err);
 
-    err = mSpake2p.ComputeRoundOne(X, &X_len);
+exit:
+    return err;
+}
+
+CHIP_ERROR SecurePairingSession::SendPBKDFParamRequest()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    System::PacketBufferHandle req;
+    uint16_t reqlen          = sizeof(PBKDFParamRequest);
+    PBKDFParamRequest * pReq = nullptr;
+
+    req = System::PacketBuffer::NewWithAvailableSize(reqlen);
+    VerifyOrExit(!req.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    pReq = reinterpret_cast<PBKDFParamRequest *>(req->Start());
+    err  = DRBG_get_bytes(pReq->mRandom, sizeof(pReq->mRandom));
     SuccessOrExit(err);
-    VerifyOrExit(CanCastTo<uint16_t>(X_len), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
-    data_len = static_cast<uint16_t>(X_len);
 
-    resp = System::PacketBuffer::NewWithAvailableSize(data_len);
+    req->SetDataLength(reqlen);
+
+    // Update commissioning hash with the pbkdf2 param request that's being sent.
+    err = mCommissioningHash.AddData(req->Start(), req->TotalLength());
+    SuccessOrExit(err);
+
+    mNextExpectedMsg = Spake2pMsgType::kPBKDFParamResponse;
+
+    err = AttachHeaderAndSend(Spake2pMsgType::kPBKDFParamRequest, req.Release_ForNow());
+    SuccessOrExit(err);
+
+    ChipLogDetail(Ble, "Sent PBKDF param request");
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        Clear();
+    }
+    return err;
+}
+
+CHIP_ERROR SecurePairingSession::HandlePBKDFParamRequest(const PacketHeader & header, const System::PacketBufferHandle & msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Request message processing
+    const uint8_t * req = msg->Start();
+    size_t reqlen       = msg->TotalLength();
+
+    VerifyOrExit(req != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(reqlen == sizeof(PBKDFParamRequest), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    ChipLogDetail(Ble, "Received PBKDF param request");
+
+    // Update commissioning hash with the received pbkdf2 param request
+    err = mCommissioningHash.AddData(req, reqlen);
+    SuccessOrExit(err);
+
+    err = SendPBKDFParamResponse();
+    SuccessOrExit(err);
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(Spake2pErrorType::kUnexpected);
+    }
+    return err;
+}
+
+CHIP_ERROR SecurePairingSession::SendPBKDFParamResponse()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    System::PacketBufferHandle resp;
+    uint16_t resplen           = 0;
+    PBKDFParamResponse * pResp = nullptr;
+
+    size_t sizeof_point = sizeof(mPoint);
+
+    VerifyOrExit(CanCastTo<uint16_t>(sizeof(PBKDFParamResponse) + mSaltLength), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    resplen = static_cast<uint16_t>(sizeof(PBKDFParamResponse) + mSaltLength);
+
+    resp = System::PacketBuffer::NewWithAvailableSize(resplen);
     VerifyOrExit(!resp.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
 
+    pResp = reinterpret_cast<PBKDFParamResponse *>(resp->Start());
+    err   = DRBG_get_bytes(pResp->mRandom, sizeof(pResp->mRandom));
```

Comment:
Please `static_assert` that the offset of `mRandom` inside `PBKDFParamResponse` is 0.

Reply:
Removed the struct.